### ISSUE TITLE
Unit test started failing because invalid version string now parsed as valid

### ISSUE THREAD

**Issue by jeffwidman** [composer/composer#11282]
I'm trying to update `composer` from `2.4.1` to `2.5.1` in Dependabot :dependabot:.

However, running into a strange unit test failure: 
* https://github.com/dependabot/dependabot-core/pull/6385#issuecomment-1401397710

The unit test relies on a fixture that has an invalid version string:
https://github.com/dependabot/dependabot-core/blob/814f7a7513a9556f23564e19901033e6c309ff58/composer/spec/fixtures/projects/invalid_version_string/composer.json#L1-L7

On composer `2.4.1`, the version string `"^3.0|4.1.x-dev as 3.0.0"` throws:
```
{"error":"Invalid version string \"^3.0|4.1.x-dev\" in \"^3.0|4.1.x-dev as 3.0.0\", the alias source must be an exact version, if it is a branch name you should prefix it with dev-"}
```

However, on `2.5.1`, it appears to parse that as a valid version string, and then fail because we're currently running `php` `7.4`:
```
{"error":"Your requirements could not be resolved to an installable set of packages.\n  Problem 1\n    - monolog\/monolog dev-main requires php >=8.1 -> your php version (7.4.33) does not satisfy that requirement.\n    - monolog\/monolog 3.x-dev is an alias of monolog\/monolog dev-main and thus requires it to be installed too.\n    - Root composer.json requires monolog\/monolog ^3.0|4.1.x-dev as 3.0.0 -> satisfiable by monolog\/monolog[3.0.0-RC1, ..., 3.x-dev (alias of dev-main)].\n"}
```

Looking at the `composer` changelog, I suspect https://github.com/composer/composer/issues/11086 / https://github.com/composer/composer/commit/8618f004f791144dc33a196a5e5346ab56f77f7b changed the parsing behavior for strings containing `|` / `||`, resulting in this version string being seen as valid.

However, the prior error indicated `"...the alias source must be an exact version, if it is a branch name you should prefix it with dev-"`, and the intent of that parsing fix doesn't mention anything about branch names, so I'm wondering if that is indeed still an invalid version string and parsing it indicates there's an upstream bug here? 

Or is that now considered a valid version string?

**Comment by Seldaek** [comment]
Yes this is now valid IMO. 4.1.x-dev is an exact version number, which is indeed also a branch but for numeric branch we use a -dev suffix instead of dev- prefix. 

The hint `if it is a branch name you should prefix it with dev-` is there because people sometimes do `feature-branch as 1.0.0` and there the fix is to use `dev-feature-branch as 1.0.0`.

**Comment by jeffwidman** [comment]
Thanks @Seldaek, I opened https://github.com/dependabot/dependabot-core/pull/6536 to fix our UT based on your clarification.
